Instance/Definition (formerly Template) (#2045)
chipsalliance/chisel3: docs/src/cookbooks/hierarchy.md

`Definition(...)` enables elaborating a module, but does not actually instantiate that module. Instead, it returns a `Definition` class which represents that module's definition.

`Instance(...)` takes a `Definition` and instantiates it, returning an `Instance` object.

Modules (classes or traits) which will be used with the `Definition`/`Instance` api should be marked with the `@instantiable` annotation at the class/trait definition.

To make a Module's members accessable from an `Instance` object, they must be annotated with the `@public` annotation.

In the following example, use `Definition`, `Instance`, `@instantiable` and `@public` to create multiple instances of one specific parameterization of a module, `AddOne`.

```scala mdoc:reset
import chisel3._
import chisel3.experimental.hierarchy._

@instantiable
class AddOne(width: Int) extends Module {
  @public val in  = IO(Input(UInt(width.W)))
  @public val out = IO(Output(UInt(width.W)))
  out := in + 1.U
}

class AddTwo(width: Int) extends Module {
  val in  = IO(Input(UInt(width.W)))
  val out = IO(Output(UInt(width.W)))
  val addOneDef = Definition(new AddOne(width))
  val i0 = Instance(addOneDef)

>   val i1 = Instance(addOneDef)

Author: Thanks!

@instantiable
class MyModule extends Module {
  @public val x = ValidType
}
```

## How do I parameterize a module by its children instances?

Prior to the introduction of this package, a parent module had to pass a child module all its parameters, when instantiating it. This had the unfortunate consequence of requiring a parent's parameters to always contain the child's parameters, which was an unnecessary coupling which lead to some anti-patterns.

Now, a parent can take a child definition as an argument, and instatiate it directly. In addition, it can analyze the parameters used in the definition create, to parameterize itself. In a sense, now the child can actually parameterize the parent.

In the following example, we create a definition of `AddOne`, and pass the definition to `AddTwo`. The width of the `AddTwo` ports are now derived from what the parameterization of the `AddOne` instances.

```scala mdoc:reset
import chisel3._
import chisel3.experimental.hierarchy._

@instantiable
class AddOne(val width: Int) extends Module {
  @public val width = width
  @public val in  = IO(Input(UInt(width.W)))
  @public val out = IO(Output(UInt(width.W)))
  out := in + 1.U
}

> class AddTwo(addOneDef: Definition[AddOne]) extends Module {

Author: I'll add this to the PR TODO list.

---
layout: docs
title:  "Hierarchy Cookbook"
section: "chisel3"
---

# Hierarchy Cookbook

* [How do I instantiate multiple instances with the same module parameterization, but avoid re-elaboration?](#how-do-i-instantiate-multiple-instances-with-the-same-module-parameterization)
* [How do I access internal fields of an instance?](#how-do-i-access-internal-fields-of-an-instance)
* [How do I make my parameters accessable from an instance?](#how-do-i-make-my-parameters-accessable-from-an-instance)
* [How do I reuse a previously elaborated module, if my new module has the same parameterization?](#how-do-i-reuse-a-previously-elaborated-module-if-my-new-module-has-the-same-parameterization)

## How do I instantiate multiple instances with the same module parameterization?

> Prior to this package, Chisel users relied on deduplication in FIRRTL to create multiple instances of the same module. This package introduces the following new APIs to enable multiply-instantiated modules directly in Chisel.

Reviewer: ```suggestion
Prior to this package, Chisel users relied on deduplication in a FIRRTL compiler to combine multiple structurally equivalent modules into one module. This package introduces the following new APIs to enable multiply-instantiated modules directly in Chisel.
```

`Definition(...)` enables elaborating a module, but does not actually instantiate that module. Instead, it returns a `Definition` class which represents that module's definition.

`Instance(...)` takes a `Definition` and instantiates it, returning an `Instance` object.

Modules (classes or traits) which will be used with the `Definition`/`Instance` api should be marked with the `@instantiable` annotation at the class/trait definition.

To make a Module's members variables accessible from an `Instance` object, they must be annotated with the `@public` annotation. Note that this is only accessible from a Scala sense -- this is not in and of itself a mechanism for cross-module references.

In the following example, use `Definition`, `Instance`, `@instantiable` and `@public` to create multiple instances of one specific parameterization of a module, `AddOne`.

```scala mdoc:reset
import chisel3._

> import chisel3.experimental.hierarchy._

Reviewer: Nit: I'd prefer to have all the things imported from `hierarchy._` explicitly spelled out. When I see wildcard imports I wonder if there is some implicit that's getting loaded or if this is just a convenience.

@instantiable
class AddOne(width: Int) extends Module {
  @public val in  = IO(Input(UInt(width.W)))
  @public val out = IO(Output(UInt(width.W)))
  out := in + 1.U
}

class AddTwo(width: Int) extends Module {
  val in  = IO(Input(UInt(width.W)))
  val out = IO(Output(UInt(width.W)))
  val addOneDef = Definition(new AddOne(width))
  val i0 = Instance(addOneDef)
  val i1 = Instance(addOneDef)
  i0.in := in
  i1.in := i0.out
  out   := i1.out
}

chisel3.stage.ChiselStage.emitVerilog(new AddTwo(10))
```

## How do I access internal fields of an instance?

> You can mark internal members of a class or trait marked with `@instantiable` with the `@public` annotation. The requirements are that the field is publically accessable, is a val or lazy val, and is a valid type.

Reviewer: ```suggestion
You can mark internal members of a class or trait marked with `@instantiable` with the `@public` annotation. The requirements are that the field is publically accessable, is a `val` or `lazy val`, and is a valid type.
```

The list of valid types are:

1. IsInstantiable
2. IsLookupable
3. Data
4. BaseModule
5. Iterable/Option containing a type that meets these requirements

> 6. Basic type like String, Int, BigInt etc.

Reviewer: Consider typewriter font for things which are actually code.

To mark a superclass's member as `@public`, use the following pattern (shown with `val clock`).

```scala mdoc:silent:reset
import chisel3._

> import chisel3.experimental.hierarchy._

Reviewer: Consider explicit import.

@instantiable
class MyModule extends Module {

>   @public val clock = clock

Reviewer: Should this use an `override`? (Does it work with an `override`?)

Reviewer 2: This is how it actually works (the macro rewrites it to make it legal Scala). I do think we should consider a better way to do it. Overriding wouldn't work (without a similar legalization rewrite) because you can't call `super` on a `val` and it would lead to null pointer exceptions when the `val` is used in the superclass.

}
```

You'll get the following error message for improperly marking something as `@public`:

```scala mdoc:reset:fail
import chisel3._

> import chisel3.experimental.hierarchy._

Reviewer: Consider explicit imports.

object NotValidType

@instantiable
class MyModule extends Module {
  @public val x = NotValidType
}
```

## How do I make my parameters accessable from an instance?

If an instance's parameters are simple (e.g. Int, String etc.) they can be marked directly with `@public`.

> Oftentimes parameters are more complicated and are contained in case classes. In these cases, mark the case class with the `IsLookupable` trait. This means that this type can be returned directly from an instance.

Reviewer: What does "returned directly from an instance"? Does this mean "accessed"? I'm being pedantic here as "return" implies something involving a function and results.

> However, ensure that these parameters are true for ALL instances of a definition. For example, if our parameters contained an id field which was instance-specific but defaulted to zero, then the definition's id would be returned for all instances. This change in behavior could lead to bugs if other code presumed the id field was correct.

Reviewer: Mega nit:

```suggestion
However, ensure that these parameters are true for _all_ instances of a definition. For example, if our parameters contained an id field which was instance-specific but defaulted to zero, then the definition's id would be returned for all instances. This change in behavior could lead to bugs if other code presumed the id field was correct.
```

Thus, it is important that when converting normal modules to use this package, you are careful about what you mark as `IsLookupable`.

> In the following example, we added the trait `IsLookupable` to allow the member to be marked `@public`.

Reviewer: This example could benefit from showing what you can do with this. E.g., accessing an instance of `MyModule`'s public `x`.

```scala mdoc:reset
import chisel3._

> import chisel3.experimental.hierarchy._

Reviewer: Consider explicit imports.

case class MyCaseClass(width: Int) extends IsLookupable

@instantiable
class MyModule extends Module {
  @public val x = MyCaseClass(10)
}
```

## How do I look up parameters from a Definition, if I don't want to instantiate it?

Just like `Instance`s, `Definition`'s also contain accessors for `@public` members. As such, you can directly
access them:

```scala mdoc:reset
import chisel3._

> import chisel3.experimental.hierarchy._

Reviewer: Consider explicit imports.
